components/service/glean/src/main/java/mozilla/components/service/glean/storages/PingStorageEngine.kt

```diff
@@ -0,0 +1,133 @@
+        private const val STORAGE_DIRECTORY = "glean/storage"
+    }
+
+    @Synchronized
+    fun store(pingType: String, pingPath: String, ping: String) {
+        val pingStorageDirectory = File(storageDirectory, pingType)
+        assertDirectory(pingStorageDirectory)
+
+        // Build the file path for the ping, using the UUID from the path for the file name
+        val pingFile = File(pingStorageDirectory, pingPath.substring(pingPath.lastIndexOf('/')))
+
+        var stream: FileOutputStream? = null
+
+        try {
+            stream = FileOutputStream(pingFile, true)
+
+            val writer = BufferedWriter(OutputStreamWriter(stream))
+            writer.write(pingPath)
+            writer.newLine()
+            writer.write(ping)
+            writer.newLine()
+            writer.flush()
+            writer.close()
+        } catch (e: IOException) {
+            logger.warn("IOException while writing event to disk", e)
+        } finally {
+            stream?.close()
+        }
+    }
+
+    fun process(pingType: String, callback: (String, String) -> Boolean): Boolean {
+        listPingFiles(pingType).forEach { file ->
+            val reader: FileReader? = null
+
+            try {
+                val bufferedReader = BufferedReader(FileReader(file))
+                val path = bufferedReader.readLine()
+                val serializedPing = bufferedReader.readLine()
+
+                val processed = serializedPing == null || callback(path, serializedPing)
+
+                if (processed) {
+                    file?.delete()
+                } else {
+                    // The callback couldn't process this file.
+                    return false
+                }
+            } catch (e: FileNotFoundException) {
+                // This shouldn't happen after we queried the directory.
+            } catch (e: IOException) {
+                // Something is not right.
+                return false
+            } finally {
+                reader?.close()
+            }
+        }
+
+        return true
+    }
+
+    @VisibleForTesting
+    internal fun listPingFiles(pingType: String): Array<File?> {
+        val pingStorageDirectory = File(storageDirectory, pingType)
+
+        val uuidPattern = Pattern.compile(FILE_PATTERN)
+
+        val uuidFilenameFilter = FilenameRegexFilter(uuidPattern)
+        return pingStorageDirectory.listFiles(uuidFilenameFilter) ?: return arrayOfNulls(0)
+    }
+
+    private fun assertDirectory(directory: File) {
+        if (!directory.exists() && !directory.mkdirs()) {
+            throw IllegalStateException(
+                "Directory doesn't exist and can't be created: " + directory.absolutePath)
+        }
+
+        if (!directory.isDirectory || !directory.canWrite()) {
+            throw IllegalStateException(
+                "Directory is not writable directory: " + directory.absolutePath)
+        }
```

Comment:
I know we've been pretty shy about exceptions.  Can you confirm my assumption that because this is called from a Worker job that all this would do is essentially cancel the worker job, but the exception wouldn't bubble up to the application?

Reply:
This actually gets called from the scheduler which runs in the client process when the ping is serialized, not the worker process.  I copied some of this from the previous telemetry ping serialization and this does bear changing to follow the 'no throw' policy of glean.